zerver/lib/actions.py

```diff
@@ -1863,18 +1873,24 @@ def notify_reaction_update(user_profile: UserProfile, message: Message,
 
 def do_add_reaction(user_profile: UserProfile, message: Message,
                     emoji_name: str, emoji_code: str, reaction_type: str) -> None:
+    """Should be called while holding a select_for_update lock
+    (e.g. via access_message(..., lock_message=True)) on the
+    Message row, to prevent IntegrityErrors if the Message is deleted
+    by another thread while this is running.
+    """
+
     reaction = Reaction(user_profile=user_profile, message=message,
                         emoji_name=emoji_name, emoji_code=emoji_code,
                         reaction_type=reaction_type)
-    try:
-        reaction.save()
-    except django.db.utils.IntegrityError:  # nocoverage
-        # This can happen when a race results in the check in views
-        # code not catching an attempt to double-add a reaction, or
-        # perhaps if the emoji_name/emoji_code mapping is busted.
-        raise JsonableError(_("Reaction already exists."))
 
-    notify_reaction_update(user_profile, message, reaction, "add")
+    reaction.save()
+
+    # We send the event after transaction is complete to make sure
+    # that an outage in RabbitMQ does not result in an outage in the
+    # database too.
```

Comment:
It's not exactly about a " database outage" - the database will be fine, but specific rows would end up being locked FOR UPDATE for absurdly long, causing certain requests to freeze etc. if they tried to acquire a FOR UPDATE (or stronger) lock on the same row.

Reply:
Makes sense. I will change that (and the commit message too).
I was also thinking if we need these comments at all? Or would be sufficient to mention this in the commit message?

Reply:
Given that this is in multiple places, I'd favor putting it in the commit message rather than copypasting the same comment. That should be our general policy - "don't send to rabbitmq while holding database locks, use transaction.on_commit'" - so this seems like something that should actually go somewhere in the documentation (after being made more precise and better phrased) - not sure where though? :thinking: Perhaps @timabbott will have more thoughts here once he gets to this PR.

Reply:
Yeah, even I was skeptical about writing the same comment everywhere, but it didn't seem trivial to use `on_commit` to me so I wrote it anyways. It'd be great if this were written in the documentation. Maybe we could also enforce this in linters?

Reply:
Yeah, I think `transaction.on_commit()` mostly speaks for itself, especially since we'll likely end up converting most `actions.py` `send_event` callers to use that pattern.